core/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -964,7 +964,7 @@ public TerminationHandle warmTopReader(final IndexShard indexShard, IndexMetaDat
             final Map<String, MappedFieldType> warmUpGlobalOrdinals = new HashMap<>();
             for (DocumentMapper docMapper : mapperService.docMappers(false)) {
                 for (FieldMapper fieldMapper : docMapper.mappers()) {
-                    final FieldDataType fieldDataType = fieldMapper.fieldType().fieldDataType();
+                    final FieldDataType fieldDataType = fieldMapper.fieldDataType();
```

Comment:
I think this would be correct then to essentially move the hack here, instead of basing on name? You could have this in a private helper method too?
```
if (fieldMapper instanceOf ParentFieldMapper) {
    handleParentField((ParentFieldMapper)fieldMapper, warmUp);
    continue;
}
```

Then that method can handle adding the field types for both the parent and/or child join fields, whichever exist?